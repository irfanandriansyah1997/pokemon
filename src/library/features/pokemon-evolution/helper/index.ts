export * from './pokemon-evolution.helper';
