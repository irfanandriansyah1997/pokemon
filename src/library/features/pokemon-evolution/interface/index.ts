export * from './pokemon-evolution.interface';
