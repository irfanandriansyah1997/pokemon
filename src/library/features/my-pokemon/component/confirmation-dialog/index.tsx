import { FC } from 'react';

import Dialog from '@/library/component/dialog';
import { IConfirmationDialog } from '@/library/features/my-pokemon/interface';
import { COLOR } from '@/library/styles/constant';
import { Text } from '@/library/styles/general.styles';
import {
  MyPokemonButtonCancel,
  MyPokemonButtonOk
} from '@/library/styles/pokemon.styles';

/**
 * Confirmation Dialog
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const ConfirmationDialog: FC<IConfirmationDialog> = ({ name, on, show }) => (
  <Dialog show={show} on={on} title="Failed">
    <Dialog.Body>
      <Text>
        Are you sure delete pokemon{` `}
        <b style={{ color: COLOR.primary, fontWeight: 500 }}>{name}</b> ?
      </Text>
    </Dialog.Body>
    <Dialog.Footer>
      <MyPokemonButtonOk
        type="button"
        onClick={() =>
          on({
            event: `on-close`
          })
        }
      >
        Cancel
      </MyPokemonButtonOk>
      <MyPokemonButtonCancel
        type="button"
        onClick={() =>
          on({
            event: `on-delete`
          })
        }
      >
        Delete
      </MyPokemonButtonCancel>
    </Dialog.Footer>
  </Dialog>
);

export default ConfirmationDialog;
