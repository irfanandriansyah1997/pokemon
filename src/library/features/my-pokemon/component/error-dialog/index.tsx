import { FC } from 'react';

import Dialog from '@/library/component/dialog';
import { IErrorDialogProps } from '@/library/features/my-pokemon/interface';
import { Text } from '@/library/styles/general.styles';

/**
 * Error Dialog
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const ErrorDialog: FC<IErrorDialogProps> = ({ on, show }) => (
  <Dialog show={show} on={on} title="Failed">
    <Dialog.Body>
      <Text>It ran away...</Text>
    </Dialog.Body>
  </Dialog>
);

export default ErrorDialog;
