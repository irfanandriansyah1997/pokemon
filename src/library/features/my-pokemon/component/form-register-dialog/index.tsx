import { verifiedIsNotEmpty } from '@99/helper';
import { ChangeEventHandler, FC, useEffect, useState } from 'react';

import Dialog from '@/library/component/dialog';
import { IFormRegisterDialog } from '@/library/features/my-pokemon/interface';
import { Text } from '@/library/styles/general.styles';
import {
  MyPokemonButtonCancel,
  MyPokemonButtonOk,
  MyPokemonTextfield
} from '@/library/styles/pokemon.styles';

/**
 * Form Registration Dialog
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const FormRegistrationDialog: FC<IFormRegisterDialog> = ({
  error: errorProps,
  on,
  show
}) => {
  const [name, setName] = useState<string>(``);
  const [error, setError] = useState(() => errorProps);

  useEffect(() => {
    setError(errorProps);
  }, [errorProps]);

  /**
   * Method Name
   * @description
   */
  const onChangeInput: ChangeEventHandler<HTMLInputElement> = ({
    target: { value }
  }): void => {
    setError(undefined);
    setName(value);
  };

  return (
    <Dialog show={show} on={on} title="Catch Pokemon Succeess!">
      <Dialog.Body>
        <MyPokemonTextfield
          type="text"
          placeholder="Give a pokemon name"
          value={name}
          onChange={onChangeInput}
        />
        <Text color="red">{error && error}</Text>
      </Dialog.Body>
      <Dialog.Footer>
        <MyPokemonButtonOk
          type="button"
          disabled={verifiedIsNotEmpty(error) || name.length === 0}
          onClick={() =>
            on({
              event: `on-submit`,
              payload: name
            })
          }
        >
          Save
        </MyPokemonButtonOk>
        <MyPokemonButtonCancel
          type="button"
          disabled={verifiedIsNotEmpty(error) || name.length === 0}
          onClick={() =>
            on({
              event: `on-close`
            })
          }
        >
          Cancel
        </MyPokemonButtonCancel>
      </Dialog.Footer>
    </Dialog>
  );
};

export default FormRegistrationDialog;
