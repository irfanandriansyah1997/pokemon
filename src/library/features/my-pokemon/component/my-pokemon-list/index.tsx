import { verifiedIsNotEmpty } from '@99/helper';
import { FC, useState } from 'react';
import Loadable from 'react-loadable';

import { useMyPokemonContext } from '@/library/features/my-pokemon/hooks/my-pokemon.hooks';
import { IPokemonDialogEvent } from '@/library/features/pokemon-detail/interface';
import PokemonCard from '@/library/features/pokemon-list/component/pokemon-card';
import { NullAble } from '@/library/interface/general';
import { IPokemon } from '@/library/interface/pokemon';
import { Text } from '@/library/styles/general.styles';
import {
  PokeListingNotFound,
  PokeListingSection
} from '@/library/styles/pokemon.styles';

const PokemonDialog = Loadable({
  loader: () =>
    import(`@/library/features/pokemon-detail/component/pokemon-dialog`),
  loading: () => null
});

/**
 * Pokemon List Component
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const MyPokemonList: FC = () => {
  const [selectedPokemon, registerPokemon] = useState<NullAble<IPokemon>>();

  const {
    state: { pokemon }
  } = useMyPokemonContext();

  /**
   * Event Listener Poke Dialog
   * @param {e} event - pokemon dialog event when user interact
   * @returns {void}
   */
  const eventListenerPokeDialog: IPokemonDialogEvent = ({ event }): void => {
    switch (event) {
      case `on-close`:
        registerPokemon(undefined);
        break;

      default:
        break;
    }
  };

  /**
   * Selected Pokemon
   * @param {string} pokemonName - pokemon name selected when user click pokemon card
   * @returns {void}
   */
  const setSelectedPokemon = (pokemonName: string): void => {
    const selected = pokemon.find(({ name }) => name === pokemonName);

    if (selected) registerPokemon(selected);
  };

  return (
    <PokeListingSection>
      <div
        style={{
          display: `flex`,
          flexDirection: `column`,
          margin: `auto`,
          maxWidth: 500,
          width: `100%`
        }}
      >
        {pokemon.map(({ ...res }) => (
          <PokemonCard key={res.id} {...res} onClick={setSelectedPokemon} />
        ))}
        {pokemon.length === 0 && (
          <PokeListingNotFound>
            <Text color="primary" fontSize="medium" fontWeight={500}>
              Pokemon
            </Text>
            <Text fontSize="text">You don`t have any pokemon</Text>
          </PokeListingNotFound>
        )}
      </div>
      <PokemonDialog
        pokemon={selectedPokemon}
        showDialog={verifiedIsNotEmpty(selectedPokemon)}
        on={eventListenerPokeDialog}
      />
    </PokeListingSection>
  );
};

export default MyPokemonList;
