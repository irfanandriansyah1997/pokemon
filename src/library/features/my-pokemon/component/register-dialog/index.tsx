import { verifiedIsNotEmpty } from '@99/helper';
import { FC, useState } from 'react';

import PokeBall from '@/library/component/pokeball';
import ConfirmationDialog from '@/library/features/my-pokemon/component/confirmation-dialog';
import ErrorDialog from '@/library/features/my-pokemon/component/error-dialog';
import FormRegistrationDialog from '@/library/features/my-pokemon/component/form-register-dialog';
import { getPokemonByCustomName } from '@/library/features/my-pokemon/helper';
import { useMyPokemonContext } from '@/library/features/my-pokemon/hooks/my-pokemon.hooks';
import {
  IConfirmationEvent,
  IFormRegisterEvent,
  IRegisterPokemonProps
} from '@/library/features/my-pokemon/interface';
import { NullAble } from '@/library/interface/general';
import { PokemonRegisterFAB } from '@/library/styles/pokemon.styles';

/**
 * Pokemon Register Dialog
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const PokemonRegisterDialog: FC<IRegisterPokemonProps> = ({
  on,
  pokemon,
  saved,
  show
}) => {
  const {
    action: { enableToCatch, registerPokemon, releasePokemon }
  } = useMyPokemonContext();
  const [loading, setLoading] = useState(false);
  const [formDialog, setFormDialog] = useState(false);
  const [errorDialog, setErrorDialog] = useState(false);
  const [confirmationDialog, setConfirmationDialog] = useState(false);
  const [errorForm, setErrorFormDialog] = useState<NullAble<string>>();

  if (!pokemon) return null;

  /**
   * Save Pokemon Into Localstorage
   * @returns {void}
   */
  const onSavePokemon = (pokemonName: string): void => {
    if (pokemon) {
      const response = registerPokemon(pokemon, pokemonName);

      if (verifiedIsNotEmpty(response))
        on({
          event: `on-save`,
          payload: response as string
        });
    }
  };

  /**
   * Delete Pokemon Into Localstorage
   * @returns {void}
   */
  const onDeletePokemon = (): void => {
    if (pokemon) {
      const response = releasePokemon(`${pokemon.customName}`);

      if (response)
        on({
          event: `on-release`
        });
    }
  };

  /**
   * Simulate On Click Button
   * @returns {void}
   */
  const onClickButton = (): void => {
    if (pokemon) {
      if (!saved) {
        setLoading(true);
        enableToCatch().then((item) => {
          setLoading(false);

          if (item) {
            setFormDialog(true);
          } else {
            setErrorDialog(true);
          }
        });
      } else {
        setConfirmationDialog(true);
      }
    }
  };

  /**
   * Event Handler Form Registration
   * @param {IEventClose | IEventSubmit} event -  event when user interact form registration
   * @returns {void}
   */
  const eventHandlerFormRegistration: IFormRegisterEvent = ({
    event,
    payload
  }): void => {
    switch (event) {
      case `on-close`:
        setFormDialog(false);
        break;

      case `on-submit`: {
        const isExist = getPokemonByCustomName(payload as string) !== undefined;

        if (isExist) {
          setErrorFormDialog(`Duplicate Pokemon`);
        } else {
          onSavePokemon(payload as string);
          setFormDialog(false);
        }

        break;
      }

      default:
        break;
    }
  };

  /**
   * Event Handler Form Registration
   * @param {IEventClose | IConfirmationEventSubmit} event -  event when user interact form confirmation
   * @returns {void}
   */
  const eventHandlerConfirmationRegistration: IConfirmationEvent = ({
    event
  }): void => {
    switch (event) {
      case `on-close`:
        setConfirmationDialog(false);
        break;

      case `on-delete`: {
        setConfirmationDialog(false);
        onDeletePokemon();
        break;
      }

      default:
        break;
    }
  };

  return (
    <>
      <PokemonRegisterFAB
        active={saved}
        loading={loading}
        show={show}
        onClick={onClickButton}
      >
        <PokeBall />
      </PokemonRegisterFAB>
      <FormRegistrationDialog
        on={eventHandlerFormRegistration}
        show={formDialog}
        error={errorForm}
      />
      <ErrorDialog on={() => setErrorDialog(false)} show={errorDialog} />
      <ConfirmationDialog
        on={eventHandlerConfirmationRegistration}
        name={pokemon?.customName}
        show={confirmationDialog}
      />
    </>
  );
};

export default PokemonRegisterDialog;
