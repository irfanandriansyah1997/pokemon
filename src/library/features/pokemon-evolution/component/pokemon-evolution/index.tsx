import { useQuery } from '@apollo/client';
import { FC, useState } from 'react';

import NextArrow from '@/assets/images/next-arrow.svg';
import { QueryEvolutionChainArgs } from '@/contract/graphql';
import {
  generateEvolution,
  getIDFromEvolution
} from '@/library/features/pokemon-evolution/helper';
import {
  IEvolutionItem,
  IEvolutionItemLabel,
  IEvolutionItemPokemon,
  IEvolutionListPokemon
} from '@/library/features/pokemon-evolution/interface';
import { IPokemon, IPokemonEvoChain } from '@/library/interface/pokemon';
import { POKEMON_EVOLUTION_QUERY } from '@/library/query';
import { Grid, Heading, Text } from '@/library/styles/general.styles';
import { PokeEvoLabel, PokeInfoItem } from '@/library/styles/pokemon.styles';

/**
 * Pokemon Evolution Label
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const PokemonEvolutionLabel: FC<IEvolutionItemLabel> = ({
  attributes,
  name
}) => (
  <PokeEvoLabel>
    <img loading="lazy" src={NextArrow} alt="evolution arrow" />
    <Text fontWeight={500} fontSize="normal" color="primary">
      {name}
    </Text>
    <Text fontWeight={400} fontSize="text" color="secondary">
      {attributes}
    </Text>
  </PokeEvoLabel>
);

/**
 * Pokemon Evolution Info
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const PokemonEvolutionInfo: FC<IEvolutionItemPokemon> = ({
  images,
  pokeName
}) => (
  <PokeInfoItem>
    <img loading="lazy" src={images} alt={pokeName} />
    <Text fontWeight={500} fontSize="text">
      {pokeName}
    </Text>
  </PokeInfoItem>
);

/**
 * Pokemon Evolution List
 * @author Irfan Andriansyah <irfan@99.co>
 * @description
 * @since 2021.08.03
 */
export const PokemonEvolutionList: FC<IEvolutionListPokemon> = ({
  evolution
}) => (
  <>
    {evolution.map(({ imageAfter, imageBefore, key, label }) => (
      <Grid templates={[`115px`, `auto`, `115px`]} key={key}>
        <PokemonEvolutionInfo {...imageBefore} />
        <PokemonEvolutionLabel {...label} />
        <PokemonEvolutionInfo {...imageAfter} />
      </Grid>
    ))}
  </>
);

/**
 * Pokemon Evolution
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.08.01
 */
const PokemonEvolution: FC<IPokemon> = ({
  name,
  pokeSpecies: { evolution_chain = {} } = {}
}) => {
  const urlID = getIDFromEvolution(evolution_chain.url || ``);
  const [evolution, setEvolution] = useState<IEvolutionItem[]>([]);

  const { loading } = useQuery<IPokemonEvoChain, QueryEvolutionChainArgs>(
    POKEMON_EVOLUTION_QUERY,
    {
      onCompleted: (response) => {
        setEvolution(generateEvolution(response));
      },
      skip: !urlID,
      variables: { id: urlID }
    }
  );

  if (loading)
    return (
      <Text fontWeight={500} fontSize="medium" textAlign="center">
        <br />
        Loading
      </Text>
    );

  return (
    <>
      <Heading>{`Evolution ${name}`}</Heading>
      <PokemonEvolutionList evolution={evolution} />
    </>
  );
};

export default PokemonEvolution;
