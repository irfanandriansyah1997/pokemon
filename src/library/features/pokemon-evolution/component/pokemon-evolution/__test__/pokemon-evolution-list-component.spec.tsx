import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { PokemonEvolutionList } from '@/library/features/pokemon-evolution/component/pokemon-evolution';

import PokemonEvoList from './__fixture__/pokemon-evo-list.fixture.json';

describe(`Testing Pokemon Evolution List`, () => {
  describe(`Testing With Snapshot`, () => {
    it(`Testing Shallow`, () => {
      const ui = shallow(<PokemonEvolutionList evolution={PokemonEvoList} />);

      expect(toJson(ui)).toMatchSnapshot();
    });

    it(`Testing Mounted`, () => {
      const ui = mount(<PokemonEvolutionList evolution={PokemonEvoList} />);

      expect(ui.html()).toMatchSnapshot();
    });
  });
});
