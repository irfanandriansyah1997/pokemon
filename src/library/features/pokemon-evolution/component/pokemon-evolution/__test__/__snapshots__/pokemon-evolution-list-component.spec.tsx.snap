// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Pokemon Evolution List Testing With Snapshot Testing Mounted 1`] = `"<div class=\\"css-2gx4tk\\"><div class=\\"css-177dxd8\\"><img loading=\\"lazy\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png\\" alt=\\"bulbasaur\\"><p font-weight=\\"500\\" font-size=\\"text\\" class=\\"css-17tivp9\\">bulbasaur</p></div><div class=\\"css-xr71pc\\"><img loading=\\"lazy\\" src=\\"next-arrow.svg\\" alt=\\"evolution arrow\\"><p font-weight=\\"500\\" font-size=\\"normal\\" color=\\"primary\\" class=\\"css-nftjgg\\">level-up</p><p font-weight=\\"400\\" font-size=\\"text\\" color=\\"secondary\\" class=\\"css-1brd6hq\\">Min Level 16</p></div><div class=\\"css-177dxd8\\"><img loading=\\"lazy\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png\\" alt=\\"ivysaur\\"><p font-weight=\\"500\\" font-size=\\"text\\" class=\\"css-17tivp9\\">ivysaur</p></div></div><div class=\\"css-2gx4tk\\"><div class=\\"css-177dxd8\\"><img loading=\\"lazy\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png\\" alt=\\"ivysaur\\"><p font-weight=\\"500\\" font-size=\\"text\\" class=\\"css-17tivp9\\">ivysaur</p></div><div class=\\"css-xr71pc\\"><img loading=\\"lazy\\" src=\\"next-arrow.svg\\" alt=\\"evolution arrow\\"><p font-weight=\\"500\\" font-size=\\"normal\\" color=\\"primary\\" class=\\"css-nftjgg\\">level-up</p><p font-weight=\\"400\\" font-size=\\"text\\" color=\\"secondary\\" class=\\"css-1brd6hq\\">Min Level 32</p></div><div class=\\"css-177dxd8\\"><img loading=\\"lazy\\" src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png\\" alt=\\"venusaur\\"><p font-weight=\\"500\\" font-size=\\"text\\" class=\\"css-17tivp9\\">venusaur</p></div></div>"`;

exports[`Testing Pokemon Evolution List Testing With Snapshot Testing Shallow 1`] = `
<Fragment>
  <Styled(div)
    key="level-0-0"
    templates={
      Array [
        "115px",
        "auto",
        "115px",
      ]
    }
  >
    <PokemonEvolutionInfo
      images="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
      pokeName="bulbasaur"
    />
    <PokemonEvolutionLabel
      attributes="Min Level 16"
      name="level-up"
    />
    <PokemonEvolutionInfo
      images="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png"
      pokeName="ivysaur"
    />
  </Styled(div)>
  <Styled(div)
    key="level-1-0"
    templates={
      Array [
        "115px",
        "auto",
        "115px",
      ]
    }
  >
    <PokemonEvolutionInfo
      images="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/2.png"
      pokeName="ivysaur"
    />
    <PokemonEvolutionLabel
      attributes="Min Level 32"
      name="level-up"
    />
    <PokemonEvolutionInfo
      images="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/3.png"
      pokeName="venusaur"
    />
  </Styled(div)>
</Fragment>
`;
